### ISSUE TITLE
While using `measureInWindow`, the `y` value is off

### ISSUE THREAD

**Issue by mrzachnugent** [software-mansion/react-native-screens#2539]
### Description

Reposting issue reported here as asked: https://github.com/react-navigation/react-navigation/issues/12294

Given a `useLayoutEffect` that calls the `measureInWindow` of a ref , the `y` value is off. It seems to be off by the header height on iOS and web. For Android, I'm not sure what is off by exactly. 

When the header is transparent or not shown, the `y` value from `measureInWindow` is correct for iOS and web but it is still off on Android.

```ts
// In a screen with default options

  React.useLayoutEffect(() => {
    triggerRef.current?.measureInWindow((x, y, width, height) => {
      setTargetRect({ x, y, width, height });
    });
  }, [setTargetRect]);

 // ...

  <Pressable ref={triggerRef}
  // ...
  
  // This should place it right below the trigger
  <View 
    style={{
    // ...
      position: "absolute",
      top: targetRect.y + targetRect.height, 
    }}
  // ...
```

### Steps to reproduce

1. Clone the minimal reproduction repo: https://github.com/mrzachnugent/react-navigation-measure-in-window-repro
2. Run `npm install && npm run dev` to install the dependencies and start Expo Go in an iOS simulator.
3. Press the "Press here to open" button, the red square should appear right below the button.
4. Try with web by pressing `w` in the terminal running metro
5. Try with android by pressing `a` in the terminal running metro
6. Toggle to a transparent header by running `git checkout transparent-header`, reload the apps and see the difference.

### Snack or a link to a repository

https://github.com/mrzachnugent/react-navigation-measure-in-window-repro

### Screens version

4.1.0

### React Native version

0.76.3

### Platforms

Android, iOS, Web

### JavaScript runtime

Hermes

### Workflow

Expo managed workflow

### Architecture

Fabric (New Architecture)

### Build type

None

### Device

iOS simulator

### Acknowledgements

Yes

**Comment by casperstr** [comment]
The same is also true for pageY  in `.measure()` when measuring an element inside a native stack header. (You get a large negative number)

**Comment by haeniya** [comment]
Having the same issue. `y` is always a large negative number. Has anyone found a workaround in the meantime?

**Comment by kkafar** [comment]
> Having the same issue. y is always a large negative number.

Hey, this is intended in case of items in header config, because the header is positioned above the screen, despite being its child in react element tree.

**Comment by kkafar** [comment]
As for original issue - we're aware of it, however this won't be fixed any time soon. We have to do many hacky workarounds to make pressable work on new architecture and the values you get are result of these hacks.

**Comment by mrzachnugent** [comment]
Hi @kkafar , thanks for the response.

Does that mean that you suggest to use anything else but a Pressable for this (like a TouchableOpacity for example)?

> In the React-Native docs, they suggest using a Pressable for a more "future proof" solution.
![image](https://github.com/user-attachments/assets/45272486-ffd0-4797-8745-e41625e9e4c4)

**Comment by mrzachnugent** [comment]
@kkafar Sorry, I don't think I understood the comment properly:

> As for original issue - we're aware of it, however this won't be fixed any time soon. We have to do many hacky workarounds to make pressable work on new architecture and the values you get are result of these hacks.

I understood that the issue is with `Pressable` and the new architecture but what you are saying is that it is not just with `Pressable`, it's with every component based on the shadow tree, correct?

As for the "workarounds", I'm a little confused as I understood that it "won't be fixed any time soon" and I must of missed the suggested workaround.

**Comment by haeniya** [comment]
> Hey, this is intended (and accurate) in case of items in header config, because the header is positioned above the screen, despite being its child in react element tree.

Hey @kkafar thanks for the reply.

So I have custom dropdown in the `headerRight` section and I'm doing the following to position the dropdown below the trigger icon:

```js
const toggleDropdown = () => {
    if (isVisible) return;

    triggerRef.current?.measureInWindow((x, y, width, height) => {
      const screenWidth = Dimensions.get('window').width;

      const rightPosition = screenWidth - (x + width); // Align dropdown's right edge with the trigger's right edge

      setDropdownPosition({
        y: y + height,
        right: rightPosition,
      });

      setIsVisible(true);
    });
  };
```
This works perfectly fine with `react-native-screens@3.31.1` & `@react-navigation/native@6.0.2` but it doesn't work anymore with `react-native-screens@4.9.1` & `@react-navigation/native@7.0.14` using the new architecture.

If this large negative number for `y` now is intended, could you please explain how it is calculated and how I now need to calculate the `y` value for my dropdown?

Many thanks.

**Comment by ionkorol** [comment]
Hope we get a proper solution to this soon. Here is my workaround for now 
https://github.com/KorSoftwareSolutions/ui/commit/b9cd74ed568c72d33466385413fcc65534a5a0b7

**Comment by kmichalikk** [comment]
I did some debugging and got some updates

tl;dr there are many things that contribute to the bug and there's not much that we can do about it at the moment

### iOS first

When placing the view at `button.top + button.height`, it lands too far down, and for `button.top + button.height - header.height` it overlaps.

<img width="279" height="218" alt="Image" src="https://github.com/user-attachments/assets/9f4f9316-56c6-4efc-b111-d85371bebc3f" />

<img width="300" height="183" alt="Image" src="https://github.com/user-attachments/assets/251fcc66-60b8-4789-87d9-d8b466f2acdb" />

This is because our first layout is usually incorrect due to the limitations of Yoga <-> native iOS/Android communication. The `measureInWindow` executes mostly from RN layout. At the very beginning RN layout engine doesn't know how large the header will be because native side haven't notified it about native layout results yet, so it acts as if it wasn't present and the screen goes to 0,0 (the button inside is placed accordingly). It quickly updates to the final correct values, but measurements are taken before it happens.

The incorrect values are then applied to the red square inside the screen. Here you can see the placement where `top = button.top + button.height - headerHeight`. Notice that the square on the left is placed correctly **under the assumption that the measurement took place when Yoga wasn't aware of the header and thought that the screen was at 0,0, similar to the situation on the right** and it actually was the place of the button that changed after native layout gave room to the header.

<img width="400" height="auto" alt="Image" src="https://github.com/user-attachments/assets/660e35bf-6a08-4365-904b-ea7459fcde10" />

<img width="400" height="auto" alt="Image" src="https://github.com/user-attachments/assets/444c4010-9faf-45e6-ac4e-c84a1ae8e248" />

----

By playing with @mrzachnugent 's test case I accidentally made it work correctly by adding `headerHeight` into `useLayoutEffect`'s dependency list. This caused recalculation of the layout with the header already in place.

Unfortunately, the described approach won't change, so the problem with **early** layout effect measurements will likely remain.

### Android

The issue is similar here.

But android has another problem which is not related to `screens`. The weird offset that doesn't fit the header height likely comes from the algorithm implemented in https://github.com/facebook/react-native/blob/8475dcc32a2622784508ff428281e9381ecd4d51/packages/react-native/ReactCommon/react/renderer/core/LayoutableShadowNode.cpp#L82. It iterates through parent views and adds up its offsets to get the origin. At some point the resulting origin is correct, but then, on ReactSurfaceView it "corrects" itself to a seemingly wrong value.

```cpp
auto shouldApplyTransformation = (policy.includeTransform && !isRootNode) ||
    (policy.includeViewportOffset && isRootNode);

// Move frame to the coordinate space of the current node.
resultFrame.origin += currentFrame.origin;

if (shouldApplyTransformation) {
      // If a node has a transform, we need to use the center of that node as
      // the origin of the transform when transforming its children (which
      // affects the result of transforms like `scale` and `rotate`).
      resultFrame = currentShadowNode->getTransform().applyWithCenter(
          resultFrame, currentFrame.getCenter());
}
```

This `shouldApplyTransformation` triggers transform, because in `DOM.cpp` the below flags are set to `true`.

```cpp
  auto layoutMetrics = getLayoutMetricsFromRoot(
      *currentRevision,
      *shadowNodeInCurrentRevision,
      {.includeTransform = true, .includeViewportOffset = true});
```

`includeViewportOffset = true` results in `ReactSurfaceView.viewportOffset` being queried

```kotlin
private val viewportOffset: Point
    get() {
      val locationOnScreen = IntArray(2)
      getLocationOnScreen(locationOnScreen)

      // we need to subtract visibleWindowCoords - to subtract possible window insets, split
      // screen or multi window
      val visibleWindowFrame = Rect()
      getWindowVisibleDisplayFrame(visibleWindowFrame)
      locationOnScreen[0] -= visibleWindowFrame.left
      locationOnScreen[1] -= visibleWindowFrame.top
      return Point(locationOnScreen[0], locationOnScreen[1])
    }
```

Notice the `getWindowVisibleDisplayFrame`. This returns the view dimensions accounting for system tray and bottom navigation insets, so the offset turns out to be negative, and moves our square to wrong place.

In `DOM.cpp`, the flags are hardcoded and there is no way to change them.

I don't know the rationale and what else breaks if it is changed, but if you manually edit the source (after enabling building from source in settings.gradle) to have this flag to false, things work correctly.

# The second issue from @haeniya 

This actually works for me in a simple case where I have a view outside of the navigation hierarchy, on both platforms (with modified RN on Android, as described above).

https://github.com/user-attachments/assets/7cecf558-cf86-4127-a9c2-3dc1ecbe9d98

**Comment by kmichalikk** [comment]
cc @kkafar

**Comment by kmichalikk** [comment]
For anyone interested https://github.com/facebook/react-native/pull/56056 has landed in RN and it seems to fix some of the Android-specific issues. I am able to use both `measure` and `measureInWindow` (with the "useHeaderHeight put in layout effect dependencies" workaround).

You can check nightly builds of react-native to see if your cases are fixed / improved. We won't be investigating more here.

### Before

<img width="841" height="830" alt="Image" src="https://github.com/user-attachments/assets/3b89f733-acef-48e5-9506-ef9993488e3b" />

measureInWindow:
<img width="694" height="270" alt="Image" src="https://github.com/user-attachments/assets/64f6069d-2900-469d-a537-380df9ce76a0" />

### After

<img width="835" height="882" alt="Image" src="https://github.com/user-attachments/assets/a070c99b-7f9e-4861-bff1-8fff2f27cb79" />
